lib/virtio_mmio/virtio_net.c

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2022 Wind River Systems, Inc.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * $FreeBSD$
+ */
+
+#include <openamp/open_amp.h>
+#include <openamp/virtqueue.h>
+#include <openamp/virtio.h>
+#include <openamp/virtio_mmio.h>
+#include <openamp/virtio_net.h>
+#include <metal/device.h>
+#include <metal/irq.h>
+
+#define LOG_MODULE_NAME "virtio_net"
+
+static int virtio_net_rx_refill(struct virtio_net_data *pdata)
+{
+	struct metal_list *node = metal_list_first(&pdata->rx_free_list);
+	struct virtio_net_rx_desc *desc = NULL;
+	struct virtqueue_buf vb[1] = {0};
+	int status = 0;
+
+	while (!virtqueue_full(pdata->vqin)) {
+		node = metal_list_first(&pdata->rx_free_list);
+		if (!node) {
+			VIRTIO_MMIO_ASSERT(0, "should have one descriptor per VQ buffer");
+			status = -ENOMEM;
+			break;
+		}
+
+		desc = metal_container_of(node, struct virtio_net_rx_desc, node);
+		memset(&desc->pkt->hdr, 0, sizeof(desc->pkt->hdr));
+
+		vb[0].buf = desc->pkt;
+		vb[0].len = pdata->hdrsize + NET_ETH_MTU;
+
+		status = virtqueue_add_buffer(pdata->vqin, vb, 0, 1, (void *)desc);
+		if (status != 0) {
+			VIRTIO_MMIO_ASSERT(0, "Should have one descriptor per VQ buffer.");
+			metal_list_add_tail(&pdata->rx_free_list, &desc->node);
+			break;
+		}
+	}
+
+	return status;
+}
+
+int virtio_mmio_net_init(struct virtio_device *vdev, struct virtqueue **vqs, char **vq_names,
+			 void (**cbs)(void *), void **cb_args, int vq_count)
+{
+	uint32_t devid, features;
+	struct virtqueue *vq = NULL;
+	struct virtio_net_data *dev_data = NULL;
+	int ret;
+	int i;
+
+	if (!vdev || !vdev->priv || !(((struct virtio_mmio_device *)(vdev->priv))->user_data) ||
```

Comment:
what about pass  user_data pointer as function parameter to avod dependency on MMIO, and adding a user_priv field in vdev?
an other solution is the  virtio_device struct is defined in user_data and be retrieved by the application with a container_of

Reply:
I refactored the code to store the user data in vdev->priv.